### PR TITLE
Add ImageParam, OutputImageParam to py bindings

### PR DESCRIPTION
**Pull request by steven-johnson**
Added OutputImageParam and Dimension and finished ImageParam; added limited (but incomplete) testing.

Note the change to Dimension in the core code: the issue here is that in an attempt to prevent a const-Dimension from being copied into a non-const version and later mutated, we made the copy ctor private with friend access. This is untenable for the Python bindings; machinery inside PyBind11 needs to be able to copy/move them, and declaring 'friend' access would be a nightmare. Instead, make it public, and add the constness checks via a runtime flag.

### CODE REVIEW

File: python_bindings/src/PyBuffer.cpp [hunk 1]
```diff
@@ -272,6 +276,13 @@ void define_buffer(py::module &m) {
             return Buffer<>(type, sizes, name);
         }), py::arg("type"), py::arg("sizes"), py::arg("name") = "")
 
+        // Note that this exists solely to allow you to create a Buffer with a null host ptr;
+        // this is necessary for some bounds-query operations (e.g. Func::infer_input_bounds).
+        // TODO: is this the best name?
+        .def_static("make_query", [](Type type, const std::vector<int> &sizes, const std::string &name) -> Buffer<> {
```
Comment (zvookin) [inline python_bindings/src/PyBuffer.cpp]: I think I'd spell it out `make_bounds_query`.

Comment (steven-johnson) [reply python_bindings/src/PyBuffer.cpp]: Done

**Comment by steven-johnson** [comment]
Leaving open for Andrew to comment before landing -- it occurs to me I might be able to minimize the changes to Dimension by using a subclass in the Python bindings.

**Comment by steven-johnson** [comment]
@abadams -- any concerns with this PR?

**Comment by abadams** [comment]
I wonder if this level of const-enforcement is really necessary. const is supposed to be something to assist the programmer. I don't think people will really copy const Dimensions and start mutating them. if they do, it's no different to copying a float reference or pointer type and then mutating the underlying object.

**Comment by steven-johnson** [comment]
Yeah, might well be overkill; I suspect this isn't much of an issue in practice. Happy to just de-constify entirely if we're comfortable with that.

**Pull request merged**